add configs api docs

agrahamlincoln opened the pull request:
Signed-off-by: Graham Rounds <graham@platform9.com>

---

Wrote this when exploring ruler/configs functionality and figured it would be valuable the official repo. Please correct anything incorrect and/or misleading I may have overlooked.

tomwilkie commented:
Thanks @agrahamlincoln, this LGTM!  We need a second +1 to merge though.  @gouthamve @bboreham @csmarchbanks ?

csmarchbanks commented on docs/configs-api.md:
```diff
@@ -0,0 +1,95 @@
+
+Returns and Accepts a JSON object representing a [Prometheus rule file](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules). 
+
+Example:
```
It's been quite awhile since I used this api, but I though these were wrapped with some extra data (rule file version, then a map of all the rule files). I think the relevant struct is [`configs.RulesConfig`](https://github.com/cortexproject/cortex/blob/master/pkg/configs/configs.go#L139).

csmarchbanks commented on docs/configs-api.md:
```diff
@@ -0,0 +1,95 @@
+
+Each tenant has a single Rule file and Alertmanager Configuration. A POST operation will effectively replace the existing copy with the configs provided in the request body.
+
+Although Prometheus works with YAML files by default, these HTTP API's only accept JSON. You may convert your YAML to JSON using a tool such as [yaml2json](https://github.com/bronze1man/yaml2json).
```
I remember succeeding in posting a raw YAML file at one point... that said I think using JSON makes the requests much simpler to figure out.

khaines replied:
I'm certain we post yaml files in our configuration system to this api. @rsteneteg , can you provide some comments here?

rsteneteg replied:
The part about only accepting JSON is correct, however just to be clear it's only the request to config service that needs to be in JSON format, the contents of your rule files should still be YAML if using the prometheus v2 rules format

jtlisi commented:
@csmarchbanks It does take a different input. Here is an example:
```json
{
  "new_config": {
    "format_version": "2",
    "files": {
      "example_rules.yaml": "groups:\n  - name: test\n    rules:\n    - record: test_ruler_sum_up\n      expr: sum(up)"
    }
  },
  "old_config": {
    "format_version": "2",
    "files": {
      "example_rules.yaml": ""
    }
  }
}```

rsteneteg commented on docs/configs-api.md:
```diff
@@ -0,0 +1,95 @@
+
+The configs api provides API-driven multi-tenant approach to prometheus' Rule files and Alertmanager configs.
+
+Each tenant has a single Rule file and Alertmanager Configuration. A POST operation will effectively replace the existing copy with the configs provided in the request body.
```
You can have multiple rule files

bboreham commented on docs/configs-api.md:
```diff
@@ -0,0 +1,95 @@
+# Configs API
+
+The configs api provides API-driven multi-tenant approach to prometheus' Rule files and Alertmanager configs.
```
minor point but I would add something like "backed by a database" - this doc mostly talks about files but the whole reason this code exists is because we wanted to use a database and the comparable code in Prometheus is hard-coded to only read files from a disk path.

rsteneteg commented on docs/configs-api.md:
````diff
@@ -0,0 +1,95 @@
+}
+```
+
+**Note:** Template files and Email Notifications are not supported in Cortex yet.
````
this PR adds support for template files https://github.com/cortexproject/cortex/pull/1237

agrahamlincoln commented:
Thanks for all the feedback! I've added the templates API endpoints, and updated the examples to be accurate. Let me know if there is anything else incorrect or that could use more elaboration.

rsteneteg commented:
@agrahamlincoln, I read the updated document, and it sounds like you can send a partial update with only templates or only rules etc.. but that is not correct, each post to any of the endpoints will replace the config with whatever was in the request. So you will need to post everything AM config, rules, templates in each request for all endpoints.

rsteneteg commented:
It would however be nice if you could post a partial update with only templates or rules etc, we should create a task to fix that

bboreham commented:
I think partial updates are listed as part of #619

bboreham commented:
I took the liberty of pushing to your branch some updates which I think bring it more into line with what the code implements.

Also I don't think we should include internal APIs in the same doc.

bboreham commented:
This can probably close #322

agrahamlincoln commented:
> I took the liberty of pushing to your branch some updates which I think bring it more into line with what the code implements.

Thanks! I appreciate you bringing together links to paint some color around why the behavior is what it is today.

agrahamlincoln commented:
Given that POST requests use the same payload, I consolidated them all into a section and used the extra real-estate to document expected return codes.

I'm pretty comfortable with this doc at this point - but definitely open to any more feedback

bboreham commented:
Let's merge this now, then any further updates can be a new PR.

bboreham commented on docs/configs-api.md:
```diff
@@ -0,0 +1,114 @@
+
+`id` - should be incremented every time data is updated; Cortex will use the config with the highest number.
+
+`rule_format_version` - allows compatability for tenants with config in Prometheus V1 format. Pass "1" or "2" according to which Prometheus version you want to match.
```
"compatability" is triggering a CI failure - better fix that or CI will be broken on master
